content/struct-classes.md

```diff
@@ -0,0 +1,627 @@
+
+~~~ scala
+trait Structural extends Product:
+  override def hashCode: Int =
+    productIterator.foldLeft(17)((hash, field) => 37 * (hash + field.##))
+  override def equals(other: Any): Boolean =
+    other match
+      case that: Structural if that.canEqual(this) =>
+        this.productIterator
+          .zip(that.productIterator)
+          .forall((x, y) => x == y)
+      case _ => false
+  override def toString: String =
+    productIterator.mkString(
+      s"${productPrefix}${this.getClass.getName}(",
+      ",",
+      ")"
+    )
+~~~
+
+Then, the class `User` used as an example could be implemented as follows:
+
+~~~ scala
+class User(val name: String, val age: Int) extends Structural:
+  def canEqual(other: Any): Boolean = other.isInstanceOf[User]
+  def productArity: Int = 2
+  def productElement(n: Int): Any = n match
+    case 0 => name
+    case 1 => age
+  private def copy(name: String = this.name, age: Int = this.age): User =
+    User(name, age)
+  def withName(newName: String): User = copy(name = newName)
+  def withAge(newAge: Int): User = copy(age = newAge)
+~~~
+
+And adding a new field `email` would be achieved as follows:
+
+~~~ scala
+class User(val name: String, val age: Int, val email: Option[String]) extends Structural:
+  def this(name: String, age: Int): User = this(name, age, email = None)
+  def canEqual(other: Any): Boolean = other.isInstanceOf[User]
+  def productArity: Int = 3
+  def productElement(n: Int): Any = n match
+    case 0 => name
+    case 1 => age
+    case 2 => email
+  private def copy(name: String = this.name, age: Int = this.age, email: Option[String] = this.email): User =
+    User(name, age, email)
+  def withName(newName: String): User = copy(name = newName)
+  def withAge(newAge: Int): User = copy(age = newAge)
+  def withEmail(newEmail: Option[String]): User = copy(email = newEmail)
+~~~
+
+This solution requires significantly more work for developers since they need
+to:
+- provide an implementation for the abstract methods `canEqual`, 
+  `productArity`, and `productElement`, which are all inherited from `Product`,
+- manually define and maintain a `copy` method,
+- explicitly mark the constructor parameters as `val` parameters.
+
+To alleviate these problems, we could change the language to automatically 
+synthesize (if not already defined in the program):
+- an implementation for `productArity` and `productElement` from
+  the class constructor parameters,
+- an implementation of `canEqual`.
+
+With these changes, the code to define the `User` type would be the following:
+
+~~~ scala
+class User(val name: String, val age: Int) extends Structural:
+  private def copy(name: String = this.name, age: Int = this.age): User =
+    User(name, age)
+  def withName(newName: String): User = copy(name = newName)
+  def withAge(newAge: Int): User = copy(age = newAge)
+~~~
+
+Developers would still have to manually define and maintain the `copy` method.
```

Comment:
Why can't the private `copy` method also be automatically synthesize?

Reply:
You are right that technically, the compiler can synthesize it. To be honest, I find it weird that the compiler does a special treatment to this trait. If it also synthesizes new methods (instead of “just” magically implementing the inherited methods), it crosses the line “that’s too many features for just extending a trait, it deserves a proper keyword in the language”.

Maybe that’s very subjective… We do have language features associated with marker traits: `Selectable` and `Dynamic`.